### DISCUSSION TITLE
How to switch from a normal `writable` store to `persisted`?

### DISCUSSION THREAD

**Discussion by ghost** [joshnuss/svelte-local-storage-store#234] [Q&A] [answered]
I am trying to use persistent storage for one of my Svelte stores but I don't have enough knowledge to get it working.

Below is my current store setup, help to switch it to using persisted would be greatly appreciated!!

`src/cart.ts`
```ts
export const cartItems = writable<CartItem[]>([]);

export const addToCart = (id: string) => {
    let items = get(cartItems);
    let itemPosition = items.findIndex(
        (item) => { return item.id == id }
    )

    if (itemPosition !== -1) {
        cartItems.update(() => {
            let updatedItems = items.map( (item) => {
                if (item.id === id) {
                    return { ...item, quantity: item.quantity + 1 };
                }
                return item;
            });
            return updatedItems;
        })
    } else {
        cartItems.update(() => {
            return [ ...items, { id: id, quantity: 1 } ]
        })
    }
}

export const removeFromCart = (id: string) => {
    let items = get(cartItems);
    let itemPosition = items.findIndex(
        (item) => { return item.id == id }
    )

    if (items[itemPosition]?.quantity - 1 === 0) {
        items.splice(itemPosition, 1)
    }

    cartItems.update(() => {
        let updatedItems = items.map((item) => {
            if (item.id === id) {
                return { ...item, quantity: item.quantity -1 }
            }
            return item;
        });
        return updatedItems;
    })
}
```

`src/app.d.ts`
```ts
declare class CartItem {
    id: string
    quantity: number
}
```

**Comment by joshnuss** [comment]
Hi Jasper!

No worries, it's super easy to do.

First, install the `svelte-persisted-store` NPM package with your package manager: 

```sh
pnpm install -D svelte-persisted-store
```

Then, import the package at the top of your file.

```javascript
import { persisted } from 'svelte-persisted-store'
```

Now all you have to do is switch your code on line 1 from `writable()` to `persisted()`.

**Note**: `persisted()` takes a `key` as the first parameter. This is the name that is for saving the data in local storage. You can use any name you'd like for the key.

```diff
---export const cartItems = writable<CartItem[]>([]);
+++export const cartItems = persisted<CartItem[]>('cart', []);
```

Now, if you open DevTools -> Application -> Local Storage, you should see the cart data.

I hope that helps! Let me know if you have any further questions

**Reply by ghost** [reply]
Hello @joshnuss 
Thank you so much for the fast response!
I have done what you said and now it has saved the data to local storage but when i reload the page it is not loading that data on the frontend. Aswell as that when i try to add items to my cart nothing is happening.
Loading the page with persisted storage on:
![image](https://github.com/joshnuss/svelte-persisted-store/assets/154771146/264547b8-78d2-4220-a85f-1487bd5925ea)
Local Storage:
![image](https://github.com/joshnuss/svelte-persisted-store/assets/154771146/a9f759e0-c4d8-4014-99e3-e6c42f63d1eb)

**Reply by ghost** [reply]
Here is my ProductCard aswell:
```ts
<script lang="ts">
    import {cartItems, addToCart, removeFromCart} from "../cart";
    import {get} from "svelte/store";
    let cart = get(cartItems);
    let cartItemIndex = cart.findIndex((item) => { return item.id === product.id });
    let cartProduct = cart[cartItemIndex];
    cartItems.subscribe((newCartValue) => {
        cart = newCartValue;
        cartItemIndex = cart.findIndex((item) => { return item.id === product.id });
        cartProduct = cart[cartItemIndex];
    });

    export let product : Product = {
        id: "",
        name: "",
        price: 0
    }
</script>

<div class="card">
    <header class="card-header">
        <h2 class="h2">{product.name}</h2>
    </header>
    {#if cartProduct !== undefined}
        <div class="card-body px-4">Quantity: <b>{cartProduct.quantity}</b></div>
    {/if}    
    <div class="card-body px-4">Price ${product.price}</div>
    <footer class="card-footer">
        <button class="p-2 rounded variant-glass-primary" on:click={() => addToCart(product.id)}>Add</button>
        <button class="p-2 rounded variant-glass-error" on:click={() => removeFromCart(product.id)}>Remove</button>
    </footer>
</div>
```

**Reply by joshnuss** [reply]
Would it be possible to create a github repo that reproduces the problem?
I think doing this would help isolate the issue.

**Reply by ghost** [reply]
Sure thing here you go.
https://github.com/jasper-at-windswept/svelte-persistance-storecart

**Reply by joshnuss** [reply]
Oh, I meant like to try to recreate a minimal repro.

I see you have functions relying on the prop `product`, but that value won't be defined at that point.

You might want to try moving that login inside `onMount`, so that the logic runs during the mounting of the component (the prop will be defined at that time)

**Reply by ghost** [reply]
> You might want to try moving that login inside `onMount`, so that the logic runs during the mounting of the component (the prop will be defined at that time)

Would you be able to give me an example of how i can do that?
I'm still just learning Svelte and am not completley familiar with when or how to use functions like onMount.

**Reply by joshnuss** [reply]
To use `onMount`, import the function `import { onMount } from 'svelte'`, then put your logic inside the `onMount` callback:

```
<!-- ProductCard.svelte -->
<script>
  import { onMount } from 'svelte'
  
  export let product
  
  onMount(() => {
    // put your logic that depends on props here
  })
</script>
```


I would highly recommend going through the Svelte tutorial, as it covers this an much much more: 
https://learn.svelte.dev/tutorial

It only takes a couple hours to go through, and will level you up quickly.

**Reply by ghost** [reply]
Oh yes sorry maybe i didn't explain properly.
I have been through the whole tutorial and a few other brief courses what I meant was I am not sure what it was you wanted me to put inside onMount?

**Reply by joshnuss** [reply]
Oh sorry, I misunderstood.

I believe the order of loading in your components is the cause of the issue.

You have code that depends on the `product` prop. That should either be inside an `onMount()` (where the prop will be available) or as a reactive statement, for example `$: loadProduct(product)`

## Option #1: onMount
 
```html
<script>
  export let product
  
  onMount(() => {
    // put logic that depends on props here
    load(product)
  })
</script>
```

## Option #2: Reactive statement

```html
<script>
  export let product
  
  // will be called whenever `product` prop changes
  $: load(product)
</script>
```

**Reply by ghost** [reply]
I think i understand a bit better what you mean now but I'm stil not sure about a few things.

1. What is the logic that depends on props that you are referring to.
2. What is the `load()` function?

**Reply by joshnuss** [reply]
> What is the logic that depends on props that you are referring to.

It's the line that is mentioned in the stack trace you provided

> What is the load() function?

It's just a name I made up, you can put whatever logic you want. For example, initializing the `cartIndex`, `cartProduct` or anything based on the props.

**Reply by ghost** [reply]
Ok so my only problem now is that the cart product variables can't be used when they are in onMount()

`productCard.svelte`
```ts
    import {onMount} from "svelte";


    onMount(() => {
        let cart = get(cartItems);
        let cartItemIndex = cart.findIndex((item) => { return item.id === product.id });
        let cartProduct = cart[cartItemIndex];
        cartItems.subscribe((newCartValue) => {
            cart = newCartValue;
            cartItemIndex = cart.findIndex((item) => { return item.id === product.id });
            cartProduct = cart[cartItemIndex];
        });
    });

    export let product : Product = {
        id: "",
        name: "",
        price: 0
    }
```

**Reply by ghost** [reply]
@joshnuss 
I'm using Skeleton UI and noticed they actually provide `localStorageStore` which was made by you.
After following the documentation there are no immediate issues after i just change my cartItems store to this:
`export const cartItems: Writable<CartItem[]> = localStorageStore('carts', []);`

The only issue now is that, It is saving and all the buttons are working but when i refresh the page it isn't loading from local storage.

**Reply by joshnuss** [reply]
Cool, glad you figured that out.

Are you using SSR? Because local storage is not available on the server. If depending on local storage, you much make sure to initialize the store on the client only.

**Reply by ghost** [reply]
I haven't specifically enabled SSR but I think it is enabled.
How can i initialize the store for the client?

**Reply by joshnuss** [reply]
Follow these instructions to disable SSR for the entire page:
https://stackoverflow.com/questions/72251017/sveltekit-disable-ssr

Or you can disable for a partial area of the page, by checking if you're in the browser before intiailizing the store:

```
<script>
  import { browser } from '$app/environment'

  const cart = browser ? persisted(...) : null
</script>

{#if cart}
 ... ok to use cart here ..
{:else}
  ...it's ssr.. show a loading message
{/if}
```

**Reply by ghost** [reply]
Ah ok thank you very much. 
Just one more thing then, From what i know won't I need the cart on all pages if for example i want to have a cart icon with the number of items in the cart.
So what parts of my pages actually "need" the cart on them?

**Reply by joshnuss** [reply]
> From what i know won't I need the cart on all pages

That a good point. In that case you should go with the second option I suggested. ie conditionaly loading the cart based on the `browser` flag.

That means the server-side rendering (SSR) can render a cart icon, just don't show the number of items. Put that in a conditional as I showed above.

> So what parts of my pages actually "need" the cart on them

Depends on your site I guess. Usually the header, product page and cart pages need it. All can be rendered server-side without and then enriched on the client.

**Reply by ghost** [reply]
So the issue is still persistent with product not being initialised. When i globally disable SSR now when the page loads with local storage i get the same 'product' error as before.
I'll have to find a way to fix that.

**Reply by ghost** [reply, marked as answer]
Good news i have resolved the issue.
In productCard.svelte i very simply needed to move `export let product` to the top of the script!

**Reply by joshnuss** [reply]
Great! Glad you got it sorted
